File: assets/static/blog_posts/algo/datastructures/deques/implementation.ex
defmodule Deque do

  defstruct [data: [], size: 0]

  def new do; %Deque{} end

  def add_front(%Deque{data: data, size: size}, item) do
    %Deque{data: data ++ [item], size: size+1}
  end

  def add_rear(%Deque{data: data, size: size}, item) do
    %Deque{data: [item | data], size: size+1}
  end

  def remove_rear(%Deque{data: []}) do; {:empty, new()} end
  def remove_rear(%Deque{data: [head | tail], size: size}) do
    {head, %Deque{data: tail, size: size-1}}
  end

  def remove_front(%Deque{data: []}) do; {:empty, new()} end
  def remove_front(%Deque{data: data, size: size}) do
    {val, data} = pop_last_elem(data)
    {val, %Deque{data: data, size: size-1}}
  end

  defp pop_last_elem(list, remaining \\ [])
  defp pop_last_elem([head], remaining) do; {head, :lists.reverse(remaining)} end
  defp pop_last_elem([head | tail], remaining) do; pop_last_elem(tail, [head | remaining]) end
end


defmodule Test do
  require Deque

  def test do
    d = Deque.new
    IO.inspect(d)

    d = Deque.add_front(d, 1)
    d = Deque.add_front(d, 2)
    d = Deque.add_front(d, 3)
    IO.inspect(d)

    d = Deque.add_rear(d, 1)
    d = Deque.add_rear(d, 2)
    d = Deque.add_rear(d, 3)
    IO.inspect(d)

    {val, d} = Deque.remove_rear(d)
    IO.inspect({val, d})
    {val, d} = Deque.remove_rear(d)
    IO.inspect({val, d})
    {val, d} = Deque.remove_rear(d)
    IO.inspect({val, d})

    {val, d} = Deque.remove_front(d)
    IO.inspect({val, d})
    {val, d} = Deque.remove_front(d)
    IO.inspect({val, d})
    {val, d} = Deque.remove_front(d)
    IO.inspect({val, d})

    {val, d} = Deque.remove_front(d)
    IO.inspect({val, d})
    {val, d} = Deque.remove_rear(d)
    IO.inspect({val, d})


  end
end

File: assets/static/blog_posts/algo/datastructures/deques/implementation.go
package main
import "fmt"

type Deque struct {
  data []interface{}
  size int
}

func (d *Deque) add_front(item interface{}) {
  d.data = append(d.data, item)
  d.size += 1
}

func (d *Deque) add_rear(item interface{}) {
  s := make([]interface{}, 1)
  s[0] = item
  d.data = append(s, d.data...)
  d.size += 1
}

func (d *Deque) remove_front() interface{} {
  if d.size == 0 {return nil}
  val := d.data[d.size-1]
  d.data = d.data[:d.size-1]
  d.size -= 1
  return val
}

func (d *Deque) remove_rear() interface{} {
  if d.size == 0 {return nil}
  val := d.data[0]
  d.data = d.data[1:]
  d.size -= 1
  return val
}

func main() {
  d := Deque{}
  d.add_front(1)
  d.add_front(2)
  d.add_front(3)
  fmt.Println(d)
  d.add_rear(-1)
  d.add_rear(-2)
  d.add_rear(-3)
  fmt.Println(d)
  d.remove_front()
  d.remove_rear()
  fmt.Println(d)
}
